packages/calling-stateful-client/src/CallAgentDeclarativeCommon.ts

```diff
@@ -0,0 +1,287 @@
+  private _callSubscribers: Map<CallCommon, CallSubscriber>;
+  private _incomingCallSubscribers: Map<string, IncomingCallSubscriber>;
+  private _declarativeIncomingCalls: Map<string, DeclarativeIncomingCall>;
+  private _declarativeCalls: Map<CallCommon, DeclarativeCallCommon>;
+  private _externalCallsUpdatedListeners: Set<CollectionUpdatedEvent<CallCommon>>;
+
+  constructor(context: CallContext, internalContext: InternalCallContext) {
+    this._context = context;
+    this._internalContext = internalContext;
+    this._callSubscribers = new Map<Call, CallSubscriber>();
+    this._incomingCallSubscribers = new Map<string, IncomingCallSubscriber>();
+    this._declarativeIncomingCalls = new Map<string, DeclarativeIncomingCall>();
+    this._declarativeCalls = new Map<Call, DeclarativeCall>();
+    this._externalCallsUpdatedListeners = new Set<CollectionUpdatedEvent<CallCommon>>();
+  }
+
+  protected abstract subscribe(): void;
+
+  // Unsubscribe is called when CallAgent is disposed. This should mean no more updating of existing call but we don't
+  // remove any existing state.
+  protected unregisterSubscriber = (): void => {
+    for (const [_, callSubscriber] of this._callSubscribers.entries()) {
+      callSubscriber.unsubscribe();
+    }
+    this._callSubscribers.clear();
+
+    for (const [_, incomingCallSubscriber] of this._incomingCallSubscribers.entries()) {
+      incomingCallSubscriber.unsubscribe();
+    }
+    this._incomingCallSubscribers.clear();
+    this._declarativeIncomingCalls.clear();
+
+    for (const [_, declarativeCall] of this._declarativeCalls.entries()) {
+      declarativeCall.unsubscribe();
+    }
+    this._declarativeCalls.clear();
+  };
+
+  protected abstract unsubscribe(): void;
+
+  protected callsUpdated = (event: {
+    added: (Call | /* @conditional-compile-remove(teams-call) */ TeamsCall)[];
+    removed: (Call | /* @conditional-compile-remove(teams-call) */ TeamsCall)[];
+  }): void => {
+    const addedStatefulCall: DeclarativeCallCommon[] = [];
+    for (const call of event.added) {
+      const statefulCall = this.addCall(call);
+      addedStatefulCall.push(statefulCall);
+    }
+    const removedStatefulCall: DeclarativeCallCommon[] = [];
+    for (const call of event.removed) {
+      disposeAllViewsFromCall(this._context, this._internalContext, call.id);
+      const callSubscriber = this._callSubscribers.get(call);
+      if (callSubscriber) {
+        callSubscriber.unsubscribe();
+        this._callSubscribers.delete(call);
+      }
+      this._context.setCallEnded(call.id, call.callEndReason);
+      const declarativeCall = this._declarativeCalls.get(call);
+      if (declarativeCall) {
+        declarativeCall.unsubscribe();
+        removedStatefulCall.push(declarativeCall);
+        this._declarativeCalls.delete(call);
+      } else {
+        if (isACSCall(call)) {
+          removedStatefulCall.push(callDeclaratify(call, this._context));
```

Comment:
Can we add `protected abstract declaratifyCall(call)` and then pass just `CallCommon` down.
You'll have to cast down in the implementation of `declaratifyCall`, but you can also add a check with `isAcsCall` there.

Reply:
```
// TeamsCallAgent

protected declaratifyCall(call: CallCommon): DeclarativeTeamsCallAgent {
    if (isACSCall(call)) {
        raise Error('UNREACHABLE');
    }
    return teamsCallDeclaratify(call as TeamsCall);
}
```